### PR TITLE
Handle saptune payload

### PR DESCRIPTION
**Pull request by CDimonaco**
# Description

Handle Saptune payload in Host Policy.

## How was this tested?

Automated tests

### CODE REVIEW

**Review by arbulu89 — COMMENTED** [review]
Hey @CDimonaco ,
Seeing the code, my major concern is this "transformation" of the payload in the policy. Not that it is done there. How it is done.
You are combining the updating the incoming message data.
Wouldn't be cleaner to get the data from the message and create a new map from scratch;
For example
```
%{packaget_version: package_version} = result = Map.get(payload, "result")
{}
|> Map.put("package_version", package_version)
|> Map.put("services", build_saptune_service(result)
|> Map.put("enabled_solutions", build_saptune_enabled_solutions(result)
    ...
```

Maybe it is a bit more verbose, but it you totally build the payload, without depending that much in the incoming data. In the current way it is really attached

File: lib/trento/application/integration/discovery/payloads/saptune_discovery_payload.ex [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+  deftype do
+    field :package_version, :string
+    field :configured_version, :string
+    field :tuning_state, :string
+    field :services, :map
+    field :notes_enabled_by_solution, {:array, :map}
+    field :notes_applied_by_solution, {:array, :map}
+    field :notes_enabled_additionally, {:array, :string}
+    field :solution_enabled, {:array, :map}
```
Comment (arbulu89) [inline lib/trento/application/integration/discovery/payloads/saptune_discovery_payload.ex]: I think `solution_enabled` should be an array of strings:
```
"Solution enabled": [
    "NETWEAVER"
],
```

File: lib/trento/application/integration/discovery/policies/host_policy.ex [hunk 2]
```diff
@@ -95,6 +140,52 @@ defmodule Trento.Integration.Discovery.HostPolicy do
            installation_source: installation_source
          })
 
+  defp build_update_saptune_command(agent_id, nil, package_version, saptune_installed),
```
Comment (arbulu89) [inline lib/trento/application/integration/discovery/policies/host_policy.ex:144]: I would put the status as last argument, to follow the same order we have in the command. Just ocd XD

File: lib/trento/domain/value_objects/saptune_staging.ex [hunk 3]
```diff
@@ -0,0 +1,11 @@
+defmodule Trento.Domain.SaptuneStaging do
+  @required_fields [:id]
```
Comment (arbulu89) [inline lib/trento/domain/value_objects/saptune_staging.ex:2]: This should be something else, `[:enabled]` maybe?

File: lib/trento/domain/value_objects/saptune_staging.ex [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+  use Trento.Type
+
+  deftype do
+    field :enabled, :boolean
+    field :notes, {:array, :string}
+    field :solutions_ids, {:array, :string}
```
Comment (arbulu89) [inline lib/trento/domain/value_objects/saptune_staging.ex:9]: I would call this `solutions`

File: lib/trento/domain/value_objects/saptune_service_status.ex [hunk 5]
```diff
@@ -0,0 +1,11 @@
+defmodule Trento.Domain.SaptuneServiceStatus do
+  @required_fields [:enabled, :active, :name]
+
+  use Trento.Type
+
+  deftype do
+    field :name, :string
+    field :enabled, :boolean
```
Comment (arbulu89) [inline lib/trento/domain/value_objects/saptune_service_status.ex:8]: I thinks `enabled/active` should be of string type, and we should forward exactly what it comes from the raw output.

File: lib/trento/application/integration/discovery/policies/host_policy.ex [hunk 6]
```diff
@@ -209,4 +300,79 @@ defmodule Trento.Integration.Discovery.HostPolicy do
...
+      "notes" => notes_staged,
+      "solutions_ids" => solutions_staged
+    }
+  end
+
+  defp format_saptune_notes(notes, additional_notes) do
+    Enum.map(notes, fn note ->
+      %{
+        "id" => note,
+        "additionally_enabled" => Enum.any?(additional_notes, &(&1 == note))
```
Comment (arbulu89) [inline lib/trento/application/integration/discovery/policies/host_policy.ex:375]: `Enum.member(additional_notes, note)`

File: lib/trento/application/integration/discovery/policies/host_policy.ex [hunk 7]
```diff
@@ -95,6 +140,52 @@ defmodule Trento.Integration.Discovery.HostPolicy do
...
+    UpdateSaptuneStatus.new(%{
+      host_id: agent_id,
+      package_version: package_version,
+      saptune_installed: saptune_installed,
+      status: %{
+        package_version: package_version,
+        configured_version: configured_version,
+        tuning_state: tuning_state,
+        services: format_saptune_services_list(services),
+        enabled_notes: format_saptune_notes(notes_enabled_additionally, notes_enabled),
```
Comment (arbulu89) [inline lib/trento/application/integration/discovery/policies/host_policy.ex:181]: Incorrect order in `enabled_notes` and `applied_notes`.
They should `(notes_enabled, notes_enabled_additionally)` (the additional notes as 2nd argument)

File: lib/trento/application/integration/discovery/policies/host_policy.ex [hunk 8]
```diff
@@ -209,4 +300,79 @@ defmodule Trento.Integration.Discovery.HostPolicy do
...
+
+  defp format_saptune_payload_keys(other_types), do: other_types
+
+  defp snake_case(payload) when is_binary(payload) do
+    payload |> String.replace(" ", "_") |> String.downcase()
+  end
+
+  defp format_saptune_services_list(%{}), do: []
+
+  defp format_saptune_services_list(service_map) do
```
Comment (arbulu89) [inline lib/trento/application/integration/discovery/policies/host_policy.ex:330]: I think this function wouldn't return a `SaptuneServiceStatus` list

**Review by arbulu89 — COMMENTED** [review]
Hey @CDimonaco,
The code looks good in general.
I dropped some comments.
Besides that, can we add `moduledocs` to all the files?

I guess once we have tests we will see if all the code is correct

File: lib/trento/application/integration/discovery/payloads/saptune_discovery_payload.ex (hunk 1 above)
Comment (CDimonaco) [reply lib/trento/application/integration/discovery/payloads/saptune_discovery_payload.ex]: done, thx for spotting

File: lib/trento/application/integration/discovery/policies/host_policy.ex (hunk 7 above)
Comment (CDimonaco) [reply lib/trento/application/integration/discovery/policies/host_policy.ex]: yes, nice catch! done

File: lib/trento/application/integration/discovery/payloads/saptune_discovery_payload.ex [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+      |> cast(attrs, fields())
+      |> cast_embed(:services)
+      |> cast_embed(:enabled_solution)
+      |> cast_embed(:applied_solution)
+      |> cast_embed(:staging)
+    end
+  end
+
+  defmodule ServiceStatus do
+    @required_fields [:enabled, :active, :name]
```
[the remark this answers is not in the document]
Comment (CDimonaco) [reply lib/trento/application/integration/discovery/payloads/saptune_discovery_payload.ex]: in our current parsing mechanism the fields will be always present with a boolean value

File: lib/trento/application/integration/discovery/policies/host_policy.ex [hunk 10]
```diff
@@ -48,6 +49,22 @@ defmodule Trento.Integration.Discovery.HostPolicy do
     end
   end
 
+  def handle(%{
+        "discovery_type" => "saptune_discovery",
+        "agent_id" => agent_id,
+        "payload" => payload
+      }) do
+    payload
+    |> Map.get("result")
```
[the remark this answers is not in the document]
Comment (CDimonaco) [reply lib/trento/application/integration/discovery/policies/host_policy.ex]: fixed

File: lib/trento/domain/value_objects/saptune_staging.ex (hunk 3 above)
Comment (CDimonaco) [reply lib/trento/domain/value_objects/saptune_staging.ex:2]: nice catch, was c&p

File: lib/trento/domain/value_objects/saptune_staging.ex (hunk 4 above)
Comment (CDimonaco) [reply lib/trento/domain/value_objects/saptune_staging.ex:9]: idk solutions in other value object refer to the whole solution object, I prefer to stick to ids, to just not confuse the stuff 👀

File: lib/trento/application/integration/discovery/policies/host_policy.ex (hunk 6 above)
Comment (CDimonaco) [reply lib/trento/application/integration/discovery/policies/host_policy.ex:388]: done thx

File: lib/trento/application/integration/discovery/policies/host_policy.ex (hunk 2 above)
Comment (CDimonaco) [reply lib/trento/application/integration/discovery/policies/host_policy.ex]: done :D

File: test/fixtures/discovery/saptune_discovery_empty_status.json [hunk 11]
```diff
@@ -0,0 +1,9 @@
+{
+    "agent_id": "9cd46919-5f19-59aa-993e-cf3736c71053",
+    "discovery_type": "saptune_discovery",
+    "payload": {
+      "saptune_installed": true,
+      "package_version": "3.1.0",
```
Comment (arbulu89) [inline test/fixtures/discovery/saptune_discovery_empty_status.json:6]: Let's put `package_version: "3.0.0"`, to simulate a real scenario

File: lib/trento/domain/value_objects/saptune_staging.ex (hunk 4 above)
Comment (arbulu89) [reply lib/trento/domain/value_objects/saptune_staging.ex:9]: well, but you have `notes` above, which is a list of note ids as well...
so we follow the same rule in both entries I guess

File: lib/trento/application/integration/discovery/policies/host_policy.ex [hunk 12]
```diff
@@ -209,4 +302,91 @@ defmodule Trento.Integration.Discovery.HostPolicy do
...
+  defp format_applied_solution(_, _), do: []
+
+  defp format_solution(%{
+         "solution_id" => solution_id,
+         "note_list" => note_list,
+         "applied_partially" => partially_applied
+       }),
+       do: %{"id" => solution_id, "notes" => note_list, "partial" => partially_applied}
+
+  defp format_saptune_staging_informations(%{
```
Comment (arbulu89) [inline lib/trento/application/integration/discovery/policies/host_policy.ex:372]: Why not simply `format_saptune_staging` ?
In any case the final `informations` with a plural `s` looks wrong

File: lib/trento/application/integration/discovery/policies/host_policy.ex [hunk 13]
```diff
@@ -209,4 +302,91 @@ defmodule Trento.Integration.Discovery.HostPolicy do
...
+    |> Enum.map(&format_saptune_payload_keys/1)
+  end
+
+  defp format_saptune_payload_keys(other_types), do: other_types
+
+  defp snake_case(payload) when is_binary(payload) do
+    payload |> String.replace(" ", "_") |> String.downcase()
+  end
+
+  defp format_saptune_services_list(services) when map_size(services) == 0, do: []
```
Comment (arbulu89) [inline lib/trento/application/integration/discovery/policies/host_policy.ex:329]: Do we need this clause?
I would expect that the next one, with the `Enum.map` with and empty `service_map` returning an empty list

**Review by arbulu89 — COMMENTED** [review]
Hey @CDimonaco , @EMaksy 
All looking great.
Once we have the `services` entry sorted out we can merge

File: lib/trento/domain/value_objects/saptune_service_status.ex (hunk 5 above)
Comment (CDimonaco) [reply lib/trento/domain/value_objects/saptune_service_status.ex]: refactored to the latest updates from the team

**Review by arbulu89 — APPROVED** [review]
Green light from my side!

File: lib/trento/application/integration/discovery/policies/host_policy.ex (hunk 13 above)
Comment (CDimonaco) [reply lib/trento/application/integration/discovery/policies/host_policy.ex]: done

File: lib/trento/domain/value_objects/saptune_note.ex [hunk 14]
```diff
@@ -0,0 +1,13 @@
+defmodule Trento.Domain.SaptuneNote do
+  @moduledoc """
+  Represents the note of Saptune.
```
Comment (arbulu89) [inline lib/trento/domain/value_objects/saptune_note.ex:3]: Represents a SAP note better.

File: lib/trento/domain/value_objects/saptune_service_status.ex [hunk 15]
```diff
@@ -0,0 +1,14 @@
+defmodule Trento.Domain.SaptuneServiceStatus do
+  @moduledoc """
+  Represents the service status of Saptune.
```
Comment (arbulu89) [inline lib/trento/domain/value_objects/saptune_service_status.ex:3]: Represents the status of a Saptune service

File: lib/trento/domain/value_objects/saptune_solution.ex [hunk 16]
```diff
@@ -0,0 +1,14 @@
+defmodule Trento.Domain.SaptuneSolution do
+  @moduledoc """
+  Represents the solutions of Saptune.
```
Comment (arbulu89) [inline lib/trento/domain/value_objects/saptune_solution.ex:3]: Represents a Saptune solution

**Review by arbulu89 — COMMENTED** [review]
@EMaksy Some suggestions on the moduledocs

**Pull request merged**